diff --git a/src/protocols/SessionLock.cpp b/src/protocols/SessionLock.cpp
index 6ff9171e303..2bec7f3cdbd 100644
--- a/src/protocols/SessionLock.cpp
+++ b/src/protocols/SessionLock.cpp
@@ -57,9 +57,13 @@ CSessionLockSurface::CSessionLockSurface(SP<CExtSessionLockSurfaceV1> resource_,
         m_surface.reset();
     });
 
-    if (m_monitor)
+    if (m_monitor) {
         PROTO::fractional->sendScale(surface_, m_monitor->m_scale);
 
+        if (m_surface)
+            m_surface->enter(m_monitor.lock());
+    }
+
     sendConfigure();
 
     m_listeners.monitorMode = m_monitor->m_events.modeChanged.listen([this] { sendConfigure(); });
