diff --git a/src/render/pass/Pass.cpp b/src/render/pass/Pass.cpp
index 3910e6a7632..5a4a8221deb 100644
--- a/src/render/pass/Pass.cpp
+++ b/src/render/pass/Pass.cpp
@@ -55,7 +55,11 @@ void CRenderPass::simplify() {
         auto opaque = el->element->opaqueRegion();
 
         if (!opaque.empty()) {
-            opaque.scale(g_pHyprOpenGL->m_renderData.pMonitor->m_scale);
+            // scale and rounding is very particular so we have to use CBoxes scale and round functions
+            if (opaque.getRects().size() == 1)
+                opaque = opaque.getExtents().scale(g_pHyprOpenGL->m_renderData.pMonitor->m_scale).round();
+            else
+                opaque.scale(g_pHyprOpenGL->m_renderData.pMonitor->m_scale);
 
             // if this intersects the liveBlur region, allow live blur to operate correctly.
             // do not occlude a border near it.
