diff --git a/src/layout/LayoutManager.cpp b/src/layout/LayoutManager.cpp
index 4a93809cdd2..3f28085f22b 100644
--- a/src/layout/LayoutManager.cpp
+++ b/src/layout/LayoutManager.cpp
@@ -95,7 +95,8 @@ void CLayoutManager::endDragTarget() {
 }
 
 void CLayoutManager::fullscreenRequestForTarget(SP<ITarget> target, eFullscreenMode currentEffectiveMode, eFullscreenMode effectiveMode) {
-    target->space()->setFullscreen(target, effectiveMode);
+    if (target->space())
+        target->space()->setFullscreen(target, effectiveMode);
 }
 
 void CLayoutManager::switchTargets(SP<ITarget> a, SP<ITarget> b, bool preserveFocus) {
diff --git a/src/layout/target/WindowGroupTarget.cpp b/src/layout/target/WindowGroupTarget.cpp
index ae883751e55..3b1dc91e97f 100644
--- a/src/layout/target/WindowGroupTarget.cpp
+++ b/src/layout/target/WindowGroupTarget.cpp
@@ -38,7 +38,8 @@ void CWindowGroupTarget::updatePos() {
 void CWindowGroupTarget::assignToSpace(const SP<CSpace>& space, std::optional<Vector2D> focalPoint) {
     ITarget::assignToSpace(space, focalPoint);
 
-    m_group->updateWorkspace(space->workspace());
+    if (space)
+        m_group->updateWorkspace(space->workspace());
 }
 
 bool CWindowGroupTarget::floating() {
