diff --git a/src/desktop/view/LayerSurface.cpp b/src/desktop/view/LayerSurface.cpp
index 1b2a3288f75..e911f590d7a 100644
--- a/src/desktop/view/LayerSurface.cpp
+++ b/src/desktop/view/LayerSurface.cpp
@@ -336,6 +336,9 @@ void CLayerSurface::onCommit() {
 
             if (m_layer == ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND || m_layer == ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM)
                 PMONITOR->m_blurFBDirty = true; // so that blur is recalc'd
+
+            if (g_pSeatManager->m_state.pointerFocus == m_wlSurface->resource())
+                g_pInputManager->simulateMouseMovement();
         }
 
         g_pHyprRenderer->arrangeLayersForMonitor(PMONITOR->m_id);
