diff --git a/src/portals/Screencopy.cpp b/src/portals/Screencopy.cpp
index 39a2bdd..14a2269 100644
--- a/src/portals/Screencopy.cpp
+++ b/src/portals/Screencopy.cpp
@@ -9,7 +9,7 @@
 #include <unistd.h>
 
 constexpr static int MAX_RETRIES        = 10;
-constexpr static int MAX_DMABUF_RETRIES = 2;
+constexpr static int MAX_DMABUF_RETRIES = 5;
 
 //
 static sdbus::Struct<std::string, uint32_t, sdbus::Variant> getFullRestoreStruct(const SSelectionData& data, uint32_t cursor) {
@@ -696,6 +696,7 @@ static void pwStreamStateChange(void* data, pw_stream_state old, pw_stream_state
     switch (state) {
         case PW_STREAM_STATE_STREAMING:
             PSTREAM->streamState = true;
+			PSTREAM->dmaBufRetries = 0;
             if (PSTREAM->pSession->sharingData.status == FRAME_NONE)
                 g_pPortalManager->m_sPortals.screencopy->startFrameCopy(PSTREAM->pSession);
             else {
