commit fcdb63e77b2761ecadba831bfc06c6adb8ce762c
Author: Zebra2711 <zebra2711@users.noreply.github.com>
Date:   Sat Apr 5 18:13:39 2025 +0700

    comctl32: fix drop combobox

diff --git a/dlls/comctl32/combo.c b/dlls/comctl32/combo.c
index 09579952f22..9ea10360d7b 100644
--- a/dlls/comctl32/combo.c
+++ b/dlls/comctl32/combo.c
@@ -22,6 +22,7 @@
 
 #include <stdarg.h>
 #include <string.h>
+#include <stdlib.h>
 
 #define OEMRESOURCE
 
@@ -914,12 +915,16 @@ static void CBUpdateEdit( LPHEADCOMBO lphc , INT index )
  */
 static void CBDropDown( LPHEADCOMBO lphc )
 {
-    HMONITOR monitor;
-    MONITORINFO mon_info;
+   HMONITOR monitor;
+   MONITORINFO mon_info;
    RECT rect,r;
    int nItems;
    int nDroppedHeight;
 
+   /* Check if running under Wayland */
+   const char *display_env = getenv("DISPLAY");
+   const char *wayland_env = getenv("WAYLAND_DISPLAY");
+
    TRACE("[%p]: drop down\n", lphc->self);
 
    CB_NOTIFY( lphc, CBN_DROPDOWN );
@@ -991,6 +996,13 @@ static void CBDropDown( LPHEADCOMBO lphc )
    mon_info.cbSize = sizeof(mon_info);
    GetMonitorInfoW( monitor, &mon_info );
 
+   if ((!display_env || !*display_env) && wayland_env && *wayland_env )
+   {
+       RECT parent_rect;
+       GetWindowRect(GetParent(lphc->self), &parent_rect);
+       mon_info.rcWork = parent_rect;
+   }
+
    if (r.bottom > mon_info.rcWork.bottom)
    {
        r.top = max( rect.top - nDroppedHeight, mon_info.rcWork.top );
