diff --git a/hyprtester/src/tests/main/dwindle.cpp b/hyprtester/src/tests/main/dwindle.cpp
index a75646c8d0e..8f17c815216 100644
--- a/hyprtester/src/tests/main/dwindle.cpp
+++ b/hyprtester/src/tests/main/dwindle.cpp
@@ -16,6 +16,7 @@ static void testFloatClamp() {
     }
 
     OK(getFromSocket("/keyword dwindle:force_split 2"));
+    OK(getFromSocket("/keyword monitor HEADLESS-2, addreserved, 0, 20, 0, 20"));
     OK(getFromSocket("/dispatch focuswindow class:c"));
     OK(getFromSocket("/dispatch setfloating class:c"));
     OK(getFromSocket("/dispatch resizewindowpixel exact 1200 900,class:c"));
@@ -24,7 +25,7 @@ static void testFloatClamp() {
 
     {
         auto str = getFromSocket("/clients");
-        EXPECT_CONTAINS(str, "at: 718,178");
+        EXPECT_CONTAINS(str, "at: 698,158");
         EXPECT_CONTAINS(str, "size: 1200,900");
     }
 
diff --git a/src/helpers/Monitor.cpp b/src/helpers/Monitor.cpp
index 74dd995b0da..50a8c1204dc 100644
--- a/src/helpers/Monitor.cpp
+++ b/src/helpers/Monitor.cpp
@@ -1509,6 +1509,10 @@ CBox CMonitor::logicalBox() {
     return {m_position, m_size};
 }
 
+CBox CMonitor::logicalBoxMinusExtents() {
+    return {m_position + m_reservedTopLeft, m_size - m_reservedTopLeft - m_reservedBottomRight};
+}
+
 void CMonitor::scheduleDone() {
     if (m_doneScheduled)
         return;
diff --git a/src/helpers/Monitor.hpp b/src/helpers/Monitor.hpp
index 5be6a7ebc9b..f1f46669813 100644
--- a/src/helpers/Monitor.hpp
+++ b/src/helpers/Monitor.hpp
@@ -298,6 +298,7 @@ class CMonitor {
     WORKSPACEID activeWorkspaceID();
     WORKSPACEID activeSpecialWorkspaceID();
     CBox        logicalBox();
+    CBox        logicalBoxMinusExtents();
     void        scheduleDone();
     uint32_t    isSolitaryBlocked(bool full = false);
     void        recheckSolitary();
diff --git a/src/layout/IHyprLayout.cpp b/src/layout/IHyprLayout.cpp
index 3886842448d..702d6ac9252 100644
--- a/src/layout/IHyprLayout.cpp
+++ b/src/layout/IHyprLayout.cpp
@@ -826,19 +826,20 @@ void IHyprLayout::fitFloatingWindowOnMonitor(PHLWINDOW w, std::optional<CBox> tb
 
     const auto EXTENTS           = w->getWindowExtentsUnified(RESERVED_EXTENTS | INPUT_EXTENTS);
     CBox       targetBoxMonLocal = tb.value_or(w->getWindowMainSurfaceBox()).translate(-PMONITOR->m_position).addExtents(EXTENTS);
+    const auto MONITOR_LOCAL_BOX = PMONITOR->logicalBoxMinusExtents().translate(-PMONITOR->m_position);
 
-    if (targetBoxMonLocal.w < PMONITOR->m_size.x) {
-        if (targetBoxMonLocal.x < 0)
-            targetBoxMonLocal.x = 0;
-        else if (targetBoxMonLocal.x + targetBoxMonLocal.w > PMONITOR->m_size.x)
-            targetBoxMonLocal.x = PMONITOR->m_size.x - targetBoxMonLocal.w;
+    if (targetBoxMonLocal.w < MONITOR_LOCAL_BOX.w) {
+        if (targetBoxMonLocal.x < MONITOR_LOCAL_BOX.x)
+            targetBoxMonLocal.x = MONITOR_LOCAL_BOX.x;
+        else if (targetBoxMonLocal.x + targetBoxMonLocal.w > MONITOR_LOCAL_BOX.w)
+            targetBoxMonLocal.x = MONITOR_LOCAL_BOX.w - targetBoxMonLocal.w;
     }
 
-    if (targetBoxMonLocal.h < PMONITOR->m_size.y) {
-        if (targetBoxMonLocal.y < 0)
-            targetBoxMonLocal.y = 0;
-        else if (targetBoxMonLocal.y + targetBoxMonLocal.h > PMONITOR->m_size.y)
-            targetBoxMonLocal.y = PMONITOR->m_size.y - targetBoxMonLocal.h;
+    if (targetBoxMonLocal.h < MONITOR_LOCAL_BOX.h) {
+        if (targetBoxMonLocal.y < MONITOR_LOCAL_BOX.y)
+            targetBoxMonLocal.y = MONITOR_LOCAL_BOX.y;
+        else if (targetBoxMonLocal.y + targetBoxMonLocal.h > MONITOR_LOCAL_BOX.h)
+            targetBoxMonLocal.y = MONITOR_LOCAL_BOX.h - targetBoxMonLocal.h;
     }
 
     *w->m_realPosition = (targetBoxMonLocal.pos() + PMONITOR->m_position + EXTENTS.topLeft).round();
