diff --git a/src/desktop/view/Window.cpp b/src/desktop/view/Window.cpp
index a0947f6718c..abb92d15f08 100644
--- a/src/desktop/view/Window.cpp
+++ b/src/desktop/view/Window.cpp
@@ -1335,6 +1335,8 @@ void CWindow::onUpdateState() {
         if (requestsID.has_value() && (requestsID.value() != MONITOR_INVALID) && !(m_suppressedEvents & SUPPRESS_FULLSCREEN_OUTPUT)) {
             if (m_isMapped) {
                 const auto monitor = g_pCompositor->getMonitorFromID(requestsID.value());
+                if (!monitor)
+                    return;
                 g_pCompositor->moveWindowToWorkspaceSafe(m_self.lock(), monitor->m_activeWorkspace);
                 Desktop::focusState()->rawMonitorFocus(monitor);
             }
@@ -2590,7 +2592,7 @@ void CWindow::commitWindow() {
             g_pHyprRenderer->damageWindow(m_self.lock());
     }
 
-    if (!m_workspace->m_visible)
+    if (!m_workspace || !m_workspace->m_visible)
         return;
 
     const auto PMONITOR = m_monitor.lock();
